From 35386aaafc834ecb3d60e882de32cf77ee331e74 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Wed, 11 Jan 2017 01:18:56 +0100
Subject: [PATCH 0749/1345] telephony: mvebu: Sort out usage of tdm2c extended
 statistics

commit  9babfc3b8f488bc00c3ce5deb107f2ec33fb942d from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

In case CONFIG_MV_TDM_EXT_STATS is enabled, while using TDMMC, checking
the statistics in procfs would result in segmentation fault. This happens
e.g. on AXP.

This commit adds following changes:
* Introduce new flag for gathering and printing extended statistics
  ('use_tdm_ext_stats'), which is a part of mv_phone_dev priv structure.
* Remove ifdefs from almost entire driver code.
* Enable the flag for TDM2C devices, depending on CONFIG_MV_TDM_EXT_STATS
  value.

Change-Id: Ife65e6264ddbc01f8d9feb0020ea130398150ada
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Signed-off-by: Igal Liberman <igall@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35870
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/telephony/mvebu_phone/mv_phone.h     |    7 +------
 drivers/telephony/mvebu_phone/mv_phone_dev.c |   23 ++++++++++++-----------
 drivers/telephony/mvebu_phone/tal/tal.h      |    2 --
 drivers/telephony/mvebu_phone/tdm2c/tdm2c.c  |    2 --
 4 files changed, 13 insertions(+), 21 deletions(-)

diff --git a/drivers/telephony/mvebu_phone/mv_phone.h b/drivers/telephony/mvebu_phone/mv_phone.h
index 47dc2d2..d71a88a 100644
--- a/drivers/telephony/mvebu_phone/mv_phone.h
+++ b/drivers/telephony/mvebu_phone/mv_phone.h
@@ -186,7 +186,6 @@ enum mv_phone_spi_mode {
 	MV_SPI_MODE_DAISY_CHAIN = 1
 };
 
-#ifdef CONFIG_MV_TDM_EXT_STATS
 struct mv_phone_extended_stats {
 	u32 int_rx_count;
 	u32 int_tx_count;
@@ -200,7 +199,6 @@ struct mv_phone_extended_stats {
 	u32 int_tx1_miss;
 	u32 pcm_restart_count;
 };
-#endif
 
 struct mv_phone_intr_info {
 	u8 *tdm_rx_buff;
@@ -260,9 +258,8 @@ struct mv_phone_dev {
 	bool pcm_stop_status;
 	int pcm_start_stop_state;
 	bool pcm_is_stopping;
-#ifdef CONFIG_MV_TDM_EXT_STATS
 	u32 pcm_stop_fail;
-#endif
+	bool use_tdm_ext_stats;
 
 	/* TDM2C SPI operation mode */
 	enum mv_phone_spi_mode tdm2c_spi_mode;
@@ -324,9 +321,7 @@ int tdm2c_init(void __iomem *base, struct device *dev,
 	       enum mv_phone_spi_mode spi_mode,
 	       bool use_pclk_external);
 int tdm2c_intr_low(struct mv_phone_intr_info *tdm_intr_info);
-#ifdef CONFIG_MV_TDM_EXT_STATS
 void tdm2c_ext_stats_get(struct mv_phone_extended_stats *tdm_ext_stats);
-#endif
 
 /* TDMMC */
 int tdmmc_init(void __iomem *base, struct device *dev, struct mv_phone_params *tdm_params,
diff --git a/drivers/telephony/mvebu_phone/mv_phone_dev.c b/drivers/telephony/mvebu_phone/mv_phone_dev.c
index e657f97..6686e0e 100644
--- a/drivers/telephony/mvebu_phone/mv_phone_dev.c
+++ b/drivers/telephony/mvebu_phone/mv_phone_dev.c
@@ -128,9 +128,7 @@
 /* Statistic printout in userspace via /proc/tdm */
 static int mv_phone_status_show(struct seq_file *m, void *v)
 {
-#ifdef CONFIG_MV_TDM_EXT_STATS
 	struct mv_phone_extended_stats tdm_ext_stats;
-#endif
 
 	seq_printf(m, "tdm_init:	%u\n", priv->tdm_init);
 	seq_printf(m, "rx_miss:		%u\n", priv->rx_miss);
@@ -138,7 +136,9 @@ static int mv_phone_status_show(struct seq_file *m, void *v)
 	seq_printf(m, "rx_over:		%u\n", priv->rx_over);
 	seq_printf(m, "tx_under:	%u\n", priv->tx_under);
 
-#ifdef CONFIG_MV_TDM_EXT_STATS
+	if (!priv->use_tdm_ext_stats)
+		return 0;
+
 	tdm2c_ext_stats_get(&tdm_ext_stats);
 
 	seq_puts(m, "\nTDM Extended Statistics:\n");
@@ -154,7 +154,7 @@ static int mv_phone_status_show(struct seq_file *m, void *v)
 	seq_printf(m, "int_tx1_miss	= %u\n", tdm_ext_stats.int_tx1_miss);
 	seq_printf(m, "pcm_restart_count= %u\n", tdm_ext_stats.pcm_restart_count);
 	seq_printf(m, "pcm_stop_fail	= %u\n", priv->pcm_stop_fail);
-#endif
+
 	return 0;
 }
 
@@ -364,9 +364,8 @@ int tdm_if_init(struct tal_params *tal_params)
 	priv->pcm_is_stopping = false;
 	priv->pcm_stop_flag = false;
 	priv->pcm_stop_status = false;
-#ifdef CONFIG_MV_TDM_EXT_STATS
 	priv->pcm_stop_fail = 0;
-#endif
+
 	/* Calculate Rx/Tx buffer size(use in callbacks) */
 	priv->buff_size = (tal_params->pcm_format * tal_params->total_lines * 80 *
 			  (tal_params->sampling_period/MV_TDM_BASE_SAMPLING_PERIOD));
@@ -524,9 +523,9 @@ static void tdm_if_stats_get(struct tal_stats *tdm_if_stats)
 	tdm_if_stats->tx_miss = priv->tx_miss;
 	tdm_if_stats->rx_over = priv->rx_over;
 	tdm_if_stats->tx_under = priv->tx_under;
-#ifdef CONFIG_MV_TDM_EXT_STATS
-	tdm2c_ext_stats_get(&tdm_if_stats->tdm_ext_stats);
-#endif
+
+	if (priv->use_tdm_ext_stats)
+		tdm2c_ext_stats_get(&tdm_if_stats->tdm_ext_stats);
 }
 
 static struct tal_if tdm2c_if = {
@@ -804,9 +803,7 @@ static void tdm2c_if_reset_channels(unsigned long arg)
 		writel(0, priv->tdm_base + CH_ENABLE_REG(1));
 		dev_warn(priv->dev, "\n%s: Channels disabling timeout (%dms)\n",
 			 __func__, MV_TDM_STOP_POLLING_TIMEOUT);
-#ifdef CONFIG_MV_TDM_EXT_STATS
 		priv->pcm_stop_fail++;
-#endif
 		mdelay(10);
 	}
 
@@ -1068,6 +1065,10 @@ static int mvebu_phone_probe(struct platform_device *pdev)
 
 		dev_info(&pdev->dev, "using %s SPI mode\n",
 			 priv->tdm2c_spi_mode ? "daisy-chain" : "direct");
+
+#ifdef CONFIG_MV_TDM_EXT_STATS
+		priv->use_tdm_ext_stats = true;
+#endif
 	}
 
 	spin_lock_init(&priv->lock);
diff --git a/drivers/telephony/mvebu_phone/tal/tal.h b/drivers/telephony/mvebu_phone/tal/tal.h
index 475c927..8fd974d 100644
--- a/drivers/telephony/mvebu_phone/tal/tal.h
+++ b/drivers/telephony/mvebu_phone/tal/tal.h
@@ -131,9 +131,7 @@ struct tal_stats {
 	u32 tx_miss;
 	u32 rx_over;
 	u32 tx_under;
-#ifdef CONFIG_MV_TDM_EXT_STATS
 	struct mv_phone_extended_stats tdm_ext_stats;
-#endif
 };
 
 struct tal_mmp_ops {
diff --git a/drivers/telephony/mvebu_phone/tdm2c/tdm2c.c b/drivers/telephony/mvebu_phone/tdm2c/tdm2c.c
index 42637f2..96674a0 100644
--- a/drivers/telephony/mvebu_phone/tdm2c/tdm2c.c
+++ b/drivers/telephony/mvebu_phone/tdm2c/tdm2c.c
@@ -985,7 +985,6 @@ void tdm2c_regs_dump(void)
 	}
 }
 
-#ifdef CONFIG_MV_TDM_EXT_STATS
 void tdm2c_ext_stats_get(struct mv_phone_extended_stats *tdm_ext_stats)
 {
 	tdm_ext_stats->int_rx_count = tdm2c->int_rx_count;
@@ -1000,7 +999,6 @@ void tdm2c_ext_stats_get(struct mv_phone_extended_stats *tdm_ext_stats)
 	tdm_ext_stats->int_tx1_miss = tdm2c->int_tx1_miss;
 	tdm_ext_stats->pcm_restart_count = tdm2c->pcm_restart_count;
 }
-#endif
 
 /* Initialize decoding windows */
 int tdm2c_set_mbus_windows(struct device *dev, void __iomem *regs,
-- 
1.7.9.5

